README.md
# cenk1cenk2/do-dyndns

[![pipeline status](https://gitlab.kilic.dev/docker/do-dyndns/badges/master/pipeline.svg)](https://gitlab.kilic.dev/docker/do-dyndns/-/commits/master) [![Docker Pulls](https://img.shields.io/docker/pulls/cenk1cenk2/do-dyndns)](https://hub.docker.com/repository/docker/cenk1cenk2/do-dyndns) [![Docker Image Size (latest by date)](https://img.shields.io/docker/image-size/cenk1cenk2/do-dyndns)](https://hub.docker.com/repository/docker/cenk1cenk2/do-dyndns) [![Docker Image Version (latest by date)](https://img.shields.io/docker/v/cenk1cenk2/do-dyndns)](https://hub.docker.com/repository/docker/cenk1cenk2/do-dyndns) [![GitHub last commit](https://img.shields.io/github/last-commit/cenk1cenk2/do-dyndns)](https://github.com/cenk1cenk2/do-dyndns)

---

## Description

This utility provides a native way to update "A" records of a domain that is managed by Digital Ocean name-servers. Simply create an API token to access and update domain records depending on the IP address of the host that is running this utility.

<!-- toc -->

- [Install](#install)
  - [Deploy with docker-compose](#deploy-with-docker-compose)
  - [Get Binary From Releases](#get-binary-from-releases)
- [Setup](#setup)
  - [Utilizing the Configuration File](#utilizing-the-configuration-file)
  - [Utilizing the Environment Variables](#utilizing-the-environment-variables)

<!-- tocstop -->

## Install

Platforms `linux-amd64`, `linux-arm64`, `linux-armv7` is supported.

### Deploy with docker-compose

Published image is available in Docker Hub as `cenk1cenk2/do-dyndns`. Will work on every platform defined.

```docker
version: "3.7"
services:
  do-dyndns:
    image: cenk1cenk2/do-dyndns
    environment:
      - DYNDNS_DOMAINS=example1.com
      - DYNDNS_SUBDOMAINS=1.example1.com,2.example2.com
      - DYNDNS_TOKEN=$DIGITAL_OCEAN_TOKEN
```

### Get Binary From Releases

You can find the natively compiled versions in the [releases](https://github.com/cenk1cenk2/do-dyndns/releases/latest).

## Setup

### Utilizing the Configuration File

Create a `.yml` file in any of the locations `.`, `/etc/do-dyndns`, `~/.config/do-dyndns` named `.do-dyndns.yml`.

You can also pass in `--config` flag to pass in the absolute path of the configuration file.

The configuration file structure is as below:

```yaml
domains:
  - example.com
subdomains:
  - 1.example.com
token: DIGITAL_OCEAN_TOKEN
repeat: 3600 # this is optional and 3600 is the default if you want repeat
```

### Utilizing the Environment Variables

To run with environment variables just pass in the variables with `DYNDNS_` prefix.

| Environment Variable | Description                                                                                    | Default |
| -------------------- | ---------------------------------------------------------------------------------------------- | ------- |
| DYNDNS_TOKEN         | Digital Ocean API token.                                                                       |         |
| DYNDNS_DOMAINS       | Comma-separated list of domains to run against.                                                |         |
| DYNDNS_SUBDOMAINS    | Comma-separated list of subdomains, which must match with the domains or else will be skipped. |         |
| DYNDNS_REPEAT        | Check repeat time in seconds.                                                                  | 3600    |


CHANGELOG.md
## [2.2.4](https://github.com/cenk1cenk2/do-dyndns/compare/v2.2.3...v2.2.4) (2022-04-19)


### Bug Fixes

* revert the order ([ceb6e90](https://github.com/cenk1cenk2/do-dyndns/commit/ceb6e90f7596f2adc0d7153c71a96665e8885f6f))

## [2.2.3](https://github.com/cenk1cenk2/do-dyndns/compare/v2.2.2...v2.2.3) (2022-04-19)


### Bug Fixes

* include tags file ([356f351](https://github.com/cenk1cenk2/do-dyndns/commit/356f35153e6a0be4e4cc59bae0eccf718b580795))
* update versioning convention ([fac54b3](https://github.com/cenk1cenk2/do-dyndns/commit/fac54b3cc5b2d8ad06748b85fec2bcb6231233f8))

## [2.2.2](https://github.com/cenk1cenk2/do-dyndns/compare/v2.2.1...v2.2.2) (2022-04-19)


### Bug Fixes

* step order ([ddb78b0](https://github.com/cenk1cenk2/do-dyndns/commit/ddb78b0331312bb353508d88a7abc702fcce5b29))

## [2.2.1](https://github.com/cenk1cenk2/do-dyndns/compare/v2.2.0...v2.2.1) (2022-04-19)


### Bug Fixes

* clean up files and version ([f1f5ba8](https://github.com/cenk1cenk2/do-dyndns/commit/f1f5ba887245ba3fbcbc6c816b82b136e6c5cb30))

# [2.2.0](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.7...v2.2.0) (2022-04-19)


### Features

* update logger and dependencies ([3bcd750](https://github.com/cenk1cenk2/do-dyndns/commit/3bcd750d38d552ba7624189e447c388b216dc6bd))

## [2.1.7](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.6...v2.1.7) (2022-04-18)


### Bug Fixes

* bump ([6ab116a](https://github.com/cenk1cenk2/do-dyndns/commit/6ab116aa171ff2b42f5d8755b381dd2b0a53287b))

## [2.1.6](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.5...v2.1.6) (2022-04-18)


### Bug Fixes

* bump ([413f173](https://github.com/cenk1cenk2/do-dyndns/commit/413f173fa13f04d4dee863d28851b0b58cf362a4))

## [2.1.5](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.4...v2.1.5) (2022-04-18)


### Bug Fixes

* update ([48844f1](https://github.com/cenk1cenk2/do-dyndns/commit/48844f1dd673c4ebe218688a886ce3fc151d214c))

## [2.1.4](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.3...v2.1.4) (2022-04-18)


### Bug Fixes

* remove drone config republish ([01362c4](https://github.com/cenk1cenk2/do-dyndns/commit/01362c45caec8259d5312c03941836006330262f))

## [2.1.3](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.2...v2.1.3) (2022-04-18)


### Bug Fixes

* use dind ([a38c131](https://github.com/cenk1cenk2/do-dyndns/commit/a38c131ed477a26ca5fd1d22f290e330337953c3))

## [2.1.2](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.1...v2.1.2) (2022-04-18)


### Bug Fixes

* update ([4ed0211](https://github.com/cenk1cenk2/do-dyndns/commit/4ed02112d65f9c253d1710ceb96328998eb98703))

## [2.1.1](https://github.com/cenk1cenk2/do-dyndns/compare/v2.1.0...v2.1.1) (2022-04-18)


### Bug Fixes

* update ([ffd23e9](https://github.com/cenk1cenk2/do-dyndns/commit/ffd23e92aeb851ea4ceedd3611ce36192031fd4d))

# [2.1.0](https://github.com/cenk1cenk2/do-dyndns/compare/v2.0.1...v2.1.0) (2022-04-18)


### Features

* update ([dce63cf](https://github.com/cenk1cenk2/do-dyndns/commit/dce63cfb012b72a198aae35975f92d17e96125b8))

# [2.0.0](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.7...v2.0.0) (2022-04-18)


### Bug Fixes

* remove gitlab configuration ([62ff1da](https://github.com/cenk1cenk2/do-dyndns/commit/62ff1da918b9ed9f1047f13a1dc1838df717cf71))
* remove versioning from git ([bfbf0f0](https://github.com/cenk1cenk2/do-dyndns/commit/bfbf0f01e900a6125ac82f5cf08842172de0031c))
* update ([867d7a6](https://github.com/cenk1cenk2/do-dyndns/commit/867d7a6918dc5b2615ce91efc009964103a793cb))


### Performance Improvements

* change the building method move to gitlab and the logger ([f8a1de2](https://github.com/cenk1cenk2/do-dyndns/commit/f8a1de2845a8970f4549f1153685faa285189bee))


### BREAKING CHANGES

* can break previous versions

# [2.0.0](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.7...v2.0.0) (2022-04-18)


### Bug Fixes

* remove gitlab configuration ([62ff1da](https://github.com/cenk1cenk2/do-dyndns/commit/62ff1da918b9ed9f1047f13a1dc1838df717cf71))


### Performance Improvements

* change the building method move to gitlab and the logger ([f8a1de2](https://github.com/cenk1cenk2/do-dyndns/commit/f8a1de2845a8970f4549f1153685faa285189bee))


### BREAKING CHANGES

* can break previous versions

## [1.2.7](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.6...v1.2.7) (2021-04-11)


### Bug Fixes

* **deps:** update module github.com/go-playground/validator/v10 to v10.5.0 ([fa3419a](https://github.com/cenk1cenk2/do-dyndns/commit/fa3419adf66551fd6fe7f2129347018b51fd8c8c))

## [1.2.6](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.5...v1.2.6) (2021-04-06)


### Bug Fixes

* **deps:** update module github.com/go-playground/validator/v10 to v10.4.2 ([596e1bc](https://github.com/cenk1cenk2/do-dyndns/commit/596e1bc42c0e8ff703469805ea9d0290669b56ec))

## [1.2.5](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.4...v1.2.5) (2021-03-09)


### Bug Fixes

* **deps:** update module github.com/antonfisher/nested-logrus-formatter to v1.3.1 ([763cd26](https://github.com/cenk1cenk2/do-dyndns/commit/763cd26dc9c835ab756f0f4d55ba54269dc27991))
* **deps:** update module github.com/sirupsen/logrus to v1.8.1 ([ed11a6f](https://github.com/cenk1cenk2/do-dyndns/commit/ed11a6fefc324df85746901e5db59ed0c987d58e))

## [1.2.4](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.3...v1.2.4) (2021-01-21)


### Bug Fixes

* delete stupid ([8a1d098](https://github.com/cenk1cenk2/do-dyndns/commit/8a1d098cd9017516ff3cb1fac7bd1cf8effe74fa))

## [1.2.3](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.2...v1.2.3) (2021-01-21)


### Bug Fixes

* update ([b703741](https://github.com/cenk1cenk2/do-dyndns/commit/b703741ece097d6e5a2ac116b3e3661e8080af2e))

## [1.2.2](https://github.com/cenk1cenk2/do-dyndns/compare/v1.2.1...v1.2.2) (2021-01-21)


### Bug Fixes

* update drone tag ([a41491f](https://github.com/cenk1cenk2/do-dyndns/commit/a41491f32510122dc6afde0e89a63c9a62dc63f0))


install.sh
#!/bin/sh

RELEASE_NAME="do-dyndns"

GITHUB_USER="cenk1cenk2"
GITHUB_REPO="do-dyndns"

RELEASE_LINUX_AMD64="linux-amd64"
CREATE_FOLDERS=

INSTALL_LOCATION_LINUX="/usr/bin"

# --------------------------------------------------------------------------------------

SEPERATOR="--------------------"
printf "[install] $RELEASE_NAME\n$SEPERATOR\n"

get_latest_release() {
  get_os_version

  LATEST_RELEASE_URL=$(
    curl --silent "https://api.github.com/repos/$GITHUB_USER/$GITHUB_USER/releases/latest" |
      grep '"browser_download_url":' |
      grep $RELEASE_PLATFORM |
      sed -E 's/.*"([^"]+)".*/\1/' |
      xargs echo
  )
}

get_os_version() {
  OS=$(uname)
  PLATFORM=$(uname -i)
  if [ $OS = "Linux" ]; then
    if [ $PLATFORM = "x86_64" ]; then
      RELEASE_PLATFORM="$RELEASE_LINUX_AMD64"
    else
      exit_unsupported_os
    fi
  else
    exit_unsupported_os
  fi
}

exit_unsupported_os() {
  echo "Unsupported operating system $OS $PLATFORM."
  exit 127
}

create_folders() {
  for i in "${CREATE_FOLDERS[@]}"; do
    eval mkdir -p ${CREATE_FOLDERS[i]}
  done
}

get_latest_release

if [ ! -z "$LATEST_RELEASE_URL" ]; then
  wget $LATEST_RELEASE_URL -O /tmp/do-dyndns
  mv /tmp/do-dyndns $INSTALL_LOCATION_LINUX/
  chmod +x $INSTALL_LOCATION_LINUX/do-dyndns

  printf "[finished] $RELEASE_NAME\n$SEPERATOR\n"
else
  printf "[failed] $RELEASE_NAME\n$SEPERATOR\n"
fi


cmd/root.go
package cmd

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"

	utils "github.com/cenk1cenk2/do-dyndns/utils"
	version "github.com/cenk1cenk2/do-dyndns/version"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
)

type iflags struct {
	once *bool
}

var flags iflags

// variables
var ip string
var ipChanged bool

// rootCmd represents the base command when called without any subcommands
var rootCmd = &cobra.Command{
	Use:     "do-dyndns",
	Short:   "Dynamically set your subdomains IP addresses that utilize Digital Ocean nameservers.",
	Example: `Please visit url for readme "https://github.com/cenk1cenk2/do-dyndns/blob/master/README.md"`,
	Version: version.Version,
	PreRun:  func(cmd *cobra.Command, args []string) { preRun(cmd, args) },
	Run:     func(cmd *cobra.Command, args []string) { run(cmd, args) },
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	if err := rootCmd.Execute(); err != nil {
		utils.Log.Fatalln(err)
		os.Exit(1)
	}
}

func preRun(cmd *cobra.Command, args []string) {
	// Load configuration
	utils.LoadConfig()
}

func run(cmd *cobra.Command, args []string) {
	// create a new set of domain and subdomain workers
	for {
		// query for ip address
		getIP()

		// if ip address changed initiate the process
		if ipChanged {

			domainWorker()

			ipChanged = false
		} else {
			utils.Log.WithFields(log.Fields{"action": "HALT"}).Debugf("Ip address has not changed.")
		}

		// break if running once
		if *flags.once {
			utils.Log.Infoln("Running once only.")
			break
		}

		// sleep for predefined time
		utils.Log.WithFields(log.Fields{"action": "HALT"}).
			Infof("Wait for %ds", utils.Config.Interval)
		time.Sleep(time.Second * time.Duration(utils.Config.Interval))
	}

	utils.Log.Infoln("Finished...")

}

func domainWorker() {
	var wg sync.WaitGroup

	// initialize variables
	var domainsProcessed []string
	var domainsProcessedMutex = &sync.Mutex{}
	var subdomainsProcessed []string
	var subdomainsProcessedMutex = &sync.Mutex{}

	// create work group
	for _, domain := range utils.Config.Domains {
		wg.Add(1)

		go func(domain string) {
			defer wg.Done()

			// check if valid domain
			re := *regexp.MustCompile(`^[^.]*\.[^.]*$`)
			if !re.MatchString(domain) {
				utils.Log.WithFields(log.Fields{"component": "DOMAIN", "action": "SKIPPED"}).
					Errorln("Not a valid domain name:", domain)

				return
			}

			utils.Log.WithFields(log.Fields{"component": "DOMAIN", "action": "STARTED"}).
				Debugln("Worker of domain:", domain)

			domainRecords, err := getDoDomainRecords(domain)

			if err != nil {
				utils.Log.WithFields(log.Fields{"component": "DO", "action": "CHECK"}).Errorln(err)

				return
			}

			// create subdomain wg
			var subdomainWG sync.WaitGroup

			// match subdomains with domains
			for _, subdomain := range utils.Config.Subdomains {
				matched, _ := regexp.MatchString(fmt.Sprintf(`\.?%s$`, domain), subdomain)

				if matched {
					subdomainWG.Add(1)

					go subdomainWorker(
						&subdomainWG,
						domain,
						subdomain,
						domainRecords,
						&subdomainsProcessed,
						subdomainsProcessedMutex,
					)
				}

			}

			utils.Log.WithFields(log.Fields{"component": "DOMAIN", "action": "FINISHED"}).
				Debugln("Worker of domain:", domain)

			subdomainWG.Wait()

			domainsProcessedMutex.Lock()
			domainsProcessed = append(domainsProcessed, domain)
			domainsProcessedMutex.Unlock()
		}(domain)
	}

	wg.Wait()
	// print out errors
	notProcessedDomains := getMissingSlice(utils.Config.Domains, domainsProcessed)
	notProcessedSubdomains := getMissingSlice(utils.Config.Subdomains, subdomainsProcessed)

	if len(notProcessedDomains) > 0 {
		utils.Log.WithFields(log.Fields{"component": "DOMAIN", "action": "FAILED"}).
			Errorf("Failed for domains: %s", strings.Join(notProcessedDomains, ", "))
	}

	if len(notProcessedSubdomains) > 0 {
		utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "FAILED"}).
			Errorf("Failed for subdomains: %s", strings.Join(notProcessedSubdomains, ", "))
	}

}

func subdomainWorker(
	wg *sync.WaitGroup,
	domain string,
	subdomain string,
	domainRecords []iDoDomainRecordsAPI,
	subdomainsProcessed *[]string,
	subdomainsProcessedMutex *sync.Mutex,
) {
	defer wg.Done()

	utils.Log.WithField("component", "DOMAIN").
		Debugf("Matched domain %s with subdomain %s\n", domain, subdomain)

	var parsedSubdomain string

	// strip subdomain to bare
	if subdomain == domain {
		// subdomain is the root domain
		parsedSubdomain = "@"

	} else {
		// subdomain is a real subdomain
		re := *regexp.MustCompile(`^(.*)\.[^.]*\.[^.]*$`)
		parsedSubdomain = re.FindStringSubmatch(subdomain)[1]

	}

	// the case where you want to change the base domain name directly
	var recordFound bool
	for _, record := range domainRecords {
		if record.Name == parsedSubdomain {
			recordFound = true

			if record.Data != ip {
				utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "STARTED"}).
					Debugln("Worker of subdomain:", domain)
				utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "CHANGE"}).
					Warningf(`Subdomain "%s" has different record of %s`, subdomain, record.Data)

				res, err := setDoDomainRecords(domain, subdomain, record.ID)

				if err != nil {
					utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "FAILED"}).
						Errorln(err)

				} else {
					utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "SUCCESS"}).Infoln(res)

				}

			} else {
				utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "SKIPPED"}).Debugln("No changes required for subodmain:", subdomain)

			}
		}
	}

	if !recordFound {
		utils.Log.WithFields(log.Fields{"component": "SUBDOMAIN", "action": "SKIPPED"}).
			Errorf(`Subdomain "%s" does not have a "A" record`, subdomain)

	} else {
		subdomainsProcessedMutex.Lock()
		*subdomainsProcessed = append(*subdomainsProcessed, subdomain)
		subdomainsProcessedMutex.Unlock()

	}
}

func getIP() {
	utils.Log.WithFields(log.Fields{"component": "IP", "action": "START"}).
		Infoln("Fetching the IP address from the API")

	body, err := createAPIRequest("GET",
		"https://api.ipify.org",
		"",
	)

	if err != nil {
		utils.Log.WithFields(log.Fields{"component": "IP", "action": "CHECK"}).Warnln(err)

		ipChanged = false
	} else {
		// check ip address
		query := string(body)
		utils.Log.WithFields(log.Fields{"component": "IP", "action": "CHECK"}).Infoln("Current IP address is:", query)

		if query != ip {
			if ip != "" {
				utils.Log.WithFields(log.Fields{"component": "IP", "action": "CHANGE"}).Warningf("IP address has been changed from %s to %s\n", ip, query)
			}

			ip = query
			ipChanged = true

		} else {
			ipChanged = false

		}

	}

}

type iGetDoDomainRecordsAPIRes struct {
	DomainRecords []iDoDomainRecordsAPI `json:"domain_records"`
	Links         struct {
	} `json:"links"`
	Meta struct {
		Total int `json:"total"`
	} `json:"meta"`
}

type iDoAPIErr struct {
	ID        string `json:"id"`
	Message   string `json:"message"`
	RequestID string `json:"request_id"`
}

type iSetDoDomainRecordsAPIRes struct {
	DomainRecord iDoDomainRecordsAPI `json:"domain_record"`
}

type iDoDomainRecordsAPI struct {
	ID       int         `json:"id"`
	Type     string      `json:"type"`
	Name     string      `json:"name"`
	Data     string      `json:"data"`
	Priority interface{} `json:"priority"`
	Port     interface{} `json:"port"`
	TTL      int         `json:"ttl"`
	Weight   interface{} `json:"weight"`
	Flags    interface{} `json:"flags"`
	Tag      interface{} `json:"tag"`
}

func getDoDomainRecords(domain string) ([]iDoDomainRecordsAPI, error) {
	body, err := createAPIRequest("GET",
		fmt.Sprintf("https://api.digitalocean.com/v2/domains/%s/records", domain),
		"",
		iRequestHeaders{Key: "content-type", Value: "application/json"},
		iRequestHeaders{Key: "Authorization", Value: fmt.Sprintf("Bearer %s", utils.Config.Token)},
	)

	if err != nil {
		return []iDoDomainRecordsAPI{}, err
	}

	var apiErr iDoAPIErr
	json.Unmarshal(body, &apiErr)

	if apiErr.ID == "not_found" {
		return []iDoDomainRecordsAPI{}, errors.New(
			fmt.Sprint("Records for the given domain can not be found: ", domain),
		)
	}

	if apiErr.ID == "Unauthorized" {
		return []iDoDomainRecordsAPI{}, errors.New(
			fmt.Sprint("Token does not seem to be valid for domain: ", domain),
		)
	}

	var value iGetDoDomainRecordsAPIRes
	json.Unmarshal(body, &value)

	// clean up records, only include A
	var domainRecords []iDoDomainRecordsAPI
	for _, record := range value.DomainRecords {
		if record.Type == "A" {
			domainRecords = append(domainRecords, record)
		}
	}

	// check the length of A records
	if len(domainRecords) == 0 {
		return []iDoDomainRecordsAPI{}, errors.New(
			fmt.Sprint("No A Records for given domain has been found: ", domain),
		)
	}

	return domainRecords, nil
}

func setDoDomainRecords(domain string, subdomain string, subdomainID int) (string, error) {
	body, err := createAPIRequest("PUT",
		fmt.Sprintf("https://api.digitalocean.com/v2/domains/%s/records/%d", domain, subdomainID),
		fmt.Sprintf(`{"data":"%s"}`, ip),
		iRequestHeaders{Key: "content-type", Value: "application/json"},
		iRequestHeaders{Key: "Authorization", Value: fmt.Sprintf("Bearer %s", utils.Config.Token)},
	)

	if err != nil {
		return "", err
	}

	var apiErr iDoAPIErr
	json.Unmarshal(body, &apiErr)

	if apiErr.ID == "not_found" {
		return "", errors.New(
			fmt.Sprint("Digital Ocean API rejected the record for subdomain:", subdomain),
		)
	}

	if apiErr.ID == "Unauthorized" {
		return "", errors.New(
			fmt.Sprint("Token does not seem to be valid for subdomain: ", subdomain),
		)
	}

	var value iSetDoDomainRecordsAPIRes
	json.Unmarshal(body, &value)

	var record = value.DomainRecord
	if record.Data != ip {
		return "", errors.New(
			fmt.Sprint("Digital Ocean API failed to set the record for subdomain:", subdomain),
		)
	}

	return fmt.Sprintln("Changed DNS record for subdomain:", subdomain), nil
}

type iRequestHeaders struct {
	Value string
	Key   string
}

func createAPIRequest(
	method string,
	url string,
	data string,
	headers ...iRequestHeaders,
) ([]byte, error) {
	// initiations
	var client = &http.Client{}

	// create request
	req, _ := http.NewRequest(method, url, bytes.NewBuffer([]byte(data)))
	req.Header.Add("User-Agent", "do-dyndns")

	for _, header := range headers {
		req.Header.Add(header.Key, header.Value)
	}

	resp, err := client.Do(req)

	if err != nil {
		return nil, errors.New(fmt.Sprint("Can not connect to the API server.", err))
	}

	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		return nil, errors.New(fmt.Sprint("Can not decode the API response.", err))
	}

	return body, nil
}

func getMissingSlice(a, b []string) []string {
	mb := make(map[string]struct{}, len(b))
	for _, x := range b {
		mb[x] = struct{}{}
	}
	var diff []string
	for _, x := range a {
		if _, found := mb[x]; !found {
			diff = append(diff, x)
		}
	}
	return diff
}

func init() {
	fmt.Println("|d|o|-|d|y|n|d|n|s|", fmt.Sprintf("%s", version.Version))

	// persistent flags
	rootCmd.PersistentFlags().
		StringVar(&utils.Cfg, "config", "", "config file ({.,/etc/do-dyndns,~/.config/do-dyndns}/.do-dyndns.yml)")
	rootCmd.PersistentFlags().
		BoolVar(&utils.LogLevelVerbose, "verbose", false, "Enable verbose logging.")

	// initialize
	cobra.OnInitialize(utils.InitiateLogger, utils.InitConfig)

	// command flags
	flags = iflags{once: rootCmd.Flags().BoolP("once", "o", false, "Run the command only once.")}
}


release.config.js
module.exports = {
  branches: ["master"],
  repositoryUrl: "https://github.com/cenk1cenk2/do-dyndns",
  plugins: [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    [
      "@google/semantic-release-replace-plugin",
      {
        replacements: [
          {
            files: ["version/v.go"],
            from: 'var Version string = "(.*)"',
            to: 'var Version string = "v${nextRelease.version}"',
            results: [
              {
                file: "version/v.go",
                hasChanged: true,
                numMatches: 1,
                numReplacements: 1,
              },
            ],
            countMatches: true,
          },
        ],
      },
    ],
    "@semantic-release/changelog",
    [
      "@semantic-release/exec",
      {
        generateNotesCmd: "echo 'latest,v${nextRelease.version}' > .tags",
      },
    ],
    [
      "@semantic-release/git",
      {
        assets: ["CHANGELOG.md", "README.md"],
      },
    ],
    [
      "@semantic-release/github",
      {
        assets: [
          {
            path: "dist/do-dyndns-linux-amd64",
            label: "do-dyndns-linux-amd64_v<%= nextRelease.version %>",
          },
          {
            path: "dist/do-dyndns-linux-armv7",
            label: "do-dyndns-linux-armv7_v<%= nextRelease.version %>",
          },
          {
            path: "dist/do-dyndns-linux-arm64",
            label: "do-dyndns-linux-arm64_v<%= nextRelease.version %>",
          },
        ],
      },
    ],
  ],
};


utils/config.go
package utils

import (
	"fmt"
	"os"
	"reflect"
	"strings"

	"github.com/go-playground/validator/v10"
	homedir "github.com/mitchellh/go-homedir"
	"github.com/spf13/viper"
)

type config struct {
	Domains    []string `mapstructure:"domains"    validate:"required,unique"`
	Subdomains []string `mapstructure:"subdomains" validate:"required,unique"`
	Token      string   `mapstructure:"token"      validate:"required"`
	Interval   int      `mapstructure:"repeat"`
}

// Cfg unparsed string config file
var Cfg string

// Config parsed config file
var Config config

// InitConfig initialize the config variable
func InitConfig() {
	// initialize viper
	if Cfg != "" {
		// Use config file from the flag.
		viper.SetConfigFile(Cfg)

	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			Log.WithField("component", "CONFIG").Fatal(err)
			os.Exit(1)
		}

		// enable config type as yml only
		viper.SetConfigType("yaml")

		// enable environment variables

		// set config and alternative config paths
		viper.AddConfigPath(".")
		viper.AddConfigPath(fmt.Sprintf("%s/.config/do-dyndns/", home))
		viper.AddConfigPath("/etc/do-dyndns/")

		// Search config in given directories with name ".do-dyndns" (without extension).
		viper.SetConfigName(".do-dyndns")
	}

	// set environment variables
	viper.SetEnvPrefix("dyndns")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))

	viper.AutomaticEnv()

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err == nil {
		Log.WithField("component", "CONFIG").Debugln("Using config file:", viper.ConfigFileUsed())
	} else {
		Log.WithField("component", "CONFIG").Debugln("Can not find config file at known locations. Trying for environment variables.")
	}
}

// LoadConfig reads in config file and ENV variables if set.
func LoadConfig() {
	err := viper.Unmarshal(&Config)
	if err != nil {
		Log.WithField("component", "CONFIG").Fatalln("Unable to decode config file, %v", err)
	}

	// set default values
	viper.SetDefault("Repeat", 3600)

	// bind environments and get the config into struct
	bindEnvs(viper.GetViper(), Config)
	viper.Unmarshal(&Config)

	// check struct
	checkConfig(Config)
}

func bindEnvs(v *viper.Viper, iface interface{}, parts ...string) {
	ifv := reflect.ValueOf(iface)
	ift := reflect.TypeOf(iface)
	for i := 0; i < ift.NumField(); i++ {
		fieldv := ifv.Field(i)
		t := ift.Field(i)
		name := strings.ToLower(t.Name)
		tag, ok := t.Tag.Lookup("mapstructure")
		if ok {
			name = tag
		}
		path := append(parts, name)
		switch fieldv.Kind() {
		case reflect.Struct:
			bindEnvs(v, fieldv.Interface(), path...)
		default:
			v.BindEnv(strings.Join(path, "."))
		}
	}
}

func checkConfig(iface config) {
	validate := validator.New()
	err := validate.Struct(iface)
	if err != nil {

		if _, ok := err.(*validator.InvalidValidationError); ok {
			Log.WithField("component", "CONFIG").Errorln(err)
			return
		}

		for _, err := range err.(validator.ValidationErrors) {
			var parsedError string

			if err.ActualTag() == "required" {
				parsedError = "is required"
			} else if err.ActualTag() == "unique" {
				parsedError = "has to be unique"
			} else {
				parsedError = fmt.Sprintf("has to be %s", err.ActualTag())
			}

			Log.WithField("component", "CONFIG").
				Errorf("%s %s and has to be %s.\n", err.Field(), parsedError, err.Type())
		}

		os.Exit(127)
	}
}
